### PR TITLE
Ignore sysctl & rlimit errors when running in UserNS

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:

/kind feature

**What this PR does / why we need it**:
Ignore sysctl errors and just print warnings instead when running in UserNS.

This is needed because sysctl is unsupported in UserNS.

Note: These sysctl parameters are set by `kubelet` itself and unrelated to `.spec.securityContext.sysctls` in Pod manifests.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:

This PR is a baby step toward integrating ["Usernetes" (Rootless Kubernetes) patch set](https://github.com/rootless-containers/usernetes/tree/master/src/patches/kubernetes) to the upstream. The patch set has been also used by [k3s](https://github.com/rancher/k3s). https://github.com/rancher/k3s/commit/cb1d94fa2731d29132703bef559cdef793793a9a https://github.com/rancher/k3s/commit/b133899935b44845be66e3c5220e05305c331b8c

~~I'll also open another PR for rootless cgroup stuff with KEP.~~

~~I didn't submit KEP for this sysct PR because the changes are really trivial and should be easy to review.~~

(**EDIT**: KEP: "Rootless mode" https://github.com/kubernetes/enhancements/pull/1371)

Aside from the rootless mode, the PR should be also useful for running Kubernetes inside LXD and maybe ChromeOS laptops.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
/sig node
/assign @johnbelamaric @vishh

cc @Random-Liu @giuseppe @ibuildthecloud

File: pkg/kubelet/cm/container_manager_linux.go
Comment (mattjmcnaughton) [inline pkg/kubelet/cm/container_manager_linux.go]: Should this log message indicate that we're running in the user namespace?

File: cmd/kube-proxy/app/server_others.go
Comment (mattjmcnaughton) [inline cmd/kube-proxy/app/server_others.go]: Were there any other options you considered beyond `rsystem` for the pkg alias? I'm wondering if we could find something slightly more descriptive.

**Review by mattjmcnaughton — COMMENTED** [review]
/ok-to-test

Just some small readability things from me :)

I'm not super familiar with this part of the code base, so interested to hear from others in sig-node as well :)

File: cmd/kube-proxy/app/server_others.go
Comment (AkihiroSuda) [reply cmd/kube-proxy/app/server_others.go]: Just followed Moby convention https://github.com/moby/moby/search?q=rsystem&unscoped_q=rsystem

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (AkihiroSuda) [reply pkg/kubelet/dockershim/docker_sandbox.go]: https://github.com/opencontainers/runc/issues/2009#issuecomment-486928621

> Checkpoint-restore is implemented entirely in CRIU, and is a very hairy process. CRIU 2.0 from a few years ago added support for unprivileged process checkpointing but it's my understanding that unprivileged process restoration is not really a thing yet.

File: cmd/kube-proxy/app/server_others.go
Comment (AkihiroSuda) [reply cmd/kube-proxy/app/server_others.go]: Let me know if you have suggestion for better alias name

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (AkihiroSuda) [reply pkg/kubelet/dockershim/docker_sandbox.go]: lol it has nothing to do with CRIU.. probably I misread the code. will update

Comment (AkihiroSuda) [reply pkg/kubelet/dockershim/docker_sandbox.go]: removed these codes

File: pkg/kubelet/cm/container_manager_linux.go
Comment (AkihiroSuda) [reply pkg/kubelet/cm/container_manager_linux.go]: fixed

File: cmd/kube-proxy/app/server_others.go
Comment (AkihiroSuda) [reply cmd/kube-proxy/app/server_others.go]: renamed to "libcontainersystem"

File: pkg/proxy/userspace/proxier.go
Comment (mattjmcnaughton) [inline pkg/proxy/userspace/proxier.go]: Why do we want this failure to return error if we are in the user namespace, but not if we aren't? It seems like, given the failure has happened, whether or not we are in the user ns should only impact the type of error message.

If we don't even want to attempt this operation if in the user ns, then we should perform the ns check earlier.

File: cmd/kube-proxy/app/server_others.go
Comment (mattjmcnaughton) [inline cmd/kube-proxy/app/server_others.go]: I like `libcontainersystem`. Its definitely longer, but I think more readable. Up to you if you prefer that or `rsystem` (I don't have any experience with the runc code base, so idk what's convention there).

**Review by mattjmcnaughton — APPROVED** [review]
/lgtm

thanks for this change :)

i'm not super familiar with usernetes, so would love another person in the community's opinion. but overall, this lgtm.

File: pkg/proxy/userspace/proxier.go
Comment (AkihiroSuda) [reply pkg/proxy/userspace/proxier.go]: > Why do we want this failure to return error if we are in the user namespace, but not if we aren't?

Maybe you have overlooked `!` symbol in the if condition?

> If we don't even want to attempt this operation if in the user ns

Worth trying even in userns, because the current process may already have > 64000 limit.

File: cmd/kube-proxy/app/server_others.go
Comment (AkihiroSuda) [reply cmd/kube-proxy/app/server_others.go]: I feel this is fine, because the package is only once used in the file.

File: pkg/proxy/userspace/proxier.go
Comment (AkihiroSuda) [reply pkg/proxy/userspace/proxier.go]: it was mistake, fixed

Comment (mattjmcnaughton) [reply pkg/proxy/userspace/proxier.go]: :+1:

**Comment by timmycarr** [comment]
Hi! I'm a 1.15 Bug triage shadow and I'd like to check in on the status of this PR. Looks like we have activity. For my tracking - Is this a release-blocking PR and is it still planned for 1.15?

**Comment by johnbelamaric** [comment]
AFAIK this is not release blocking and at this point should be targeted to 1.16.

/lgtm

**Comment by derekwaynecarr** [comment]
Can we hold accepting changes to give an impression of supporting kubelet in a userns when we have a full design that explains how it is expected to work rather than piecemeal address issues?  This seems similar to past efforts to run kubelet in a container that we had abandoned for a variety of issues.  If there is a desire to bring back a containerized kubelet (directly or indirectly), we should evaluate that in a broader KEP.  Let's discuss in SIG Node.

**Comment by AkihiroSuda** [comment]
>  when we have a full design that explains how it is expected to work rather than piecemeal address issues? 

The full design is to put the entire Kubernetes components (`kubelet`, CRI, OCI, CNI, and all `kube-*`) in UserNS.
Currently cgroup doesn't work (unless chmod/chown-ed) and we need to add "none" cgroup driver (KEP: https://github.com/kubernetes/enhancements/pull/1084), but after the migration to cgroup2, we will be able to safely support cgroup.

You can find more information in the slide decks of my joint talk with @giuseppe at FOSDEM 2019 : https://www.slideshare.net/AkihiroSuda/rootless-kubernetes
Our patched Kubernetes is available at https://github.com/rootless-containers/usernetes and also integrated to k3s.

> This seems similar to past efforts to run kubelet in a container that we had abandoned for a variety of issues.

This proposal should be rather compared to run everything (not just kubelet) in a container as in [`kind`](https://github.com/kubernetes-sigs/kind), which is still officially supported.

**Comment by derekwaynecarr** [comment]
@AkihiroSuda I appreciate the details.  I still prefer we get a complete end-to-end design on the desired end-state, and then work backwards on the incremental enhancements we can make to reach that state.

**Comment by dims** [comment]
/milestone v1.16

Let's move this out please based on the last comment from @derekwaynecarr

**Comment by AkihiroSuda** [comment]
We are discussing supporting rootless mode with cgroup2 rather than with `cgroup-driver=none` (https://github.com/kubernetes/enhancements/pull/1084), but anyway I think this PR is ready to merge regardless to the cgroup design.

**Comment by AkihiroSuda** [comment]
rebased

**Comment by AkihiroSuda** [comment]
`Test_buildClientCertificateManager` failure seems unrelated

/test pull-kubernetes-bazel-test

**Comment by AkihiroSuda** [comment]
@mattjmcnaughton @johnbelamaric  @bowei @vishh PTAL?

**Comment by mattjmcnaughton** [comment]
> @mattjmcnaughton @johnbelamaric @bowei @vishh PTAL?

Apologies if I'm wrong, but my understanding is that we're still waiting on feedback/approval from @derekwaynecarr, re his last comment?

**Comment by makoscafee** [comment]
Hello! @AkihiroSuda I'm a bug triage shadow for the 1.16 release cycle I'd like to check it's status. The code freeze is starting 29th August (about 2 weeks from now) and while there is plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.16, is it still planned for this release?

**Comment by AkihiroSuda** [comment]
@makoscafee Not likely to be planned for 1.16.

@giuseppe is planning to open a KEP for cgroup2 soon, which will address @derekwaynecarr 's concern. https://github.com/kubernetes/kubernetes/pull/78635#issuecomment-499167370

**Comment by xmudrii** [comment]
As per,

> @makoscafee Not likely to be planned for 1.16.
> 
> @giuseppe is planning to open a KEP for cgroup2 soon, which will address @derekwaynecarr 's concern. #78635 (comment)

I'm going to move this issue out of the milestone. If there are any concerns or anything changes in meanwhile, please ping me or ping SIG-Release on Slack.

/milestone clear

**Comment by AkihiroSuda** [comment]
New KEP is ready: https://github.com/kubernetes/enhancements/pull/1371

@derekwaynecarr PTAL.

Aside from rootless, the PR should be also useful for running Kubernetes inside LXD and maybe ChromeOS laptops.

**Comment by AkihiroSuda** [comment]
@derekwaynecarr PTAL? 🙏

**Comment by dims** [comment]
@derekwaynecarr @yujuhong @dchen1107 This is the PR that `k3s` already ships!! Can we please make sure this ends up in 1.18?

- https://github.com/rancher/k3s/commit/1d6a02285ecab9c5d7ae3d644371982403565b85
- https://github.com/rancher/k3s/commit/aa5098f2231860b56d3894d7d5faae3d57b529d7

Many thanks for your work @AkihiroSuda 

Thanks,
Dims

**Comment by AkihiroSuda** [comment]
rebased

**Comment by AkihiroSuda** [comment]
Any chance to get this merged?
This PR has been already cherry-picked by multiple Kubernetes distros (k3s (https://github.com/kubernetes/kubernetes/pull/78635#issuecomment-564001540) and Silverkube)

**Comment by AkihiroSuda** [comment]
I'm temporarily closing this until https://github.com/kubernetes/enhancements/pull/1370 settles, but if anybody needs this to be merged for other purposes, feel free to open PR to carry this.

Probably this PR is useful for LXC and Crostini users.

**Comment by h-vetinari** [comment]
> I'm temporarily closing this until kubernetes/enhancements#1370 settles, but if anybody needs this to be merged for other purposes, feel free to open PR to carry this.

Since that KEP has been merged a while ago, should this PR be reopened?

**Comment by AkihiroSuda** [comment]
I'll reopen PR with support for rootless cgroup, right after the release of libcontainer (runc) rc11

**Comment by AkihiroSuda** [comment]
This one is now reopened as https://github.com/kubernetes/kubernetes/pull/92863

**Pull request closed without merging**
